File: binary_indexed_tree.c

struct BIT {
    int n;
    int* nums; // index starting from 1
    int* c; // cumulative
};

int lowbit(int x) {
    return x & (-x);
}

struct BIT* BITCreate(int* a, int n) {
    struct BIT* b = malloc(sizeof(struct BIT));
    
    b->n = n;
    int size = n * sizeof(int) + 1;
    b->nums = malloc(size);
    b->c = malloc(size);

    memset((void*)b->nums, 0, size);
    memset((void*)b->c, 0, size);

    if (a == NULL) {
        return b;
    }

    // init
    for (int i = 1; i <= n; i++) {
        int x = a[i-1];
        b->nums[i] = x;
        for (int j = i; j <= n; j += lowbit(j)) {
            b->c[j] += x;
        }
    }

    return b;
}

void BITModify(struct BIT* b, int i, int value) {
    int delta = value - b->nums[i];
    b->nums[i] = value;

    for (int j = i; j <= b->n; j += lowbit(j)) {
        b->c[j] += delta;
    }
}

// sum of nums[1 ~ k]
int BITSum(struct BIT* b, int k) {
    int sum = 0;
    for (int j = k; j > 0; j -= lowbit(j)) {
        sum += b->c[j];
    }
    return sum;
}

File: binary_indexed_tree.java

import java.util.*;

class BIT {
    int[] sum;
    int[] nums;
    int n;
    
    public BIT(int[] nums) {
        n = nums.length;
        this.nums = nums; 
        this.sum = new int[n+1];
        for (int i = 0; i < n; i++) {
            changeValue(i, nums[i]);
        }
    }
    
    public void update(int i, int val) {
        int delta = val - nums[i];
        nums[i] = val;
        changeValue(i, delta);
    }

    public int sumRange(int i, int j) {
        int s = sumPrefix(j);
        int t = sumPrefix(i-1);
        return s - t;
    }
    
    //////////////////////////////////////////////////
    
    private void changeValue(int i, int delta) {
        int j = i + 1;
        while(j <= n) {
            sum[j] += delta;
            j += lowbit(j);
        }
    }
    
    private int sumPrefix(int i) {
        int s = 0;
        int j = i + 1;
        while(j > 0) {
            s += sum[j];
            j -= lowbit(j);
        }
        return s;
    }
    
    private int lowbit(int x) {
        return x & (-x);
    }
}
